File: python/dp/213. House Robber II.py
'''
dp
Time: O(n)
Space: O(1)
'''

class Solution:
    def rob(self, nums: List[int]) -> int:
        # nums[0] (edge case) : if len(nums) == 1
        return max(nums[0], self.houseRobberI(nums[1:]), self.houseRobberI(nums[:-1]))

    def houseRobberI(self, nums):
        rob1, rob2 = 0, 0
        for n in nums:
            rob1, rob2 = rob2, max(rob1 + n, rob2)

        return rob2

File: javascript/dp/213. House Robber II.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var rob = function(nums) {
    return Math.max(nums[0], houseRobberI(nums.slice(1)), houseRobberI(nums.slice(0, -1)));
};

var houseRobberI = (nums) => {
    let rob1 = 0, rob2 = 0;
    for (n of nums) {
        newRob = Math.max(rob1 + n, rob2);
        rob1 = rob2;
        rob2 = newRob;
    }
    return rob2;
}